### ISSUE TITLE
Finish API for BCO 

### ISSUE THREAD

**Issue by HadleyKing** [biocompute-objects/galaxy#4]
The API here is incomplete. Finish mapping the files and then test and review. 
https://github.com/biocompute-objects/galaxy/blob/ead0cc81eed98a6a257e36470f10bfdf176d6c86/lib/galaxy/webapps/galaxy/api/histories.py#L528-L559

**Comment by HadleyKing** [comment]
Update to this API [WIP]: 
https://github.com/biocompute-objects/galaxy/blob/a730bd4c7ad44dce5fe67acf8c12575860691f7f/lib/galaxy/webapps/galaxy/api/histories.py#L528-L614

**Comment by nsoranzo** [comment]
@HadleyKing Can you open a PR to this repo? It would be easier to see the differences.

**Comment by HadleyKing** [comment]
@nsoranzo 
A PR between which repos? I'm sorry I am still a newbie. 
Do you mean like this:
![Screenshot 2019-10-01 10 57 19](https://user-images.githubusercontent.com/20343214/65974026-3e8a5500-e43a-11e9-8589-e6bd4f3f3c12.png)
 
Or do you men the main Galaxy dev repo?

**Comment by nsoranzo** [comment]
@HadleyKing Sorry, I didn't realise that you already committed it to this repo! Nothing to be done then.

**Comment by HadleyKing** [comment]
It is mostly sudo code and I think may need some additional functions (Which I think I outlined in most of the comments), but I wanted to get the ball rolling, rather than wait for perfection.

**Comment by HadleyKing** [comment]
I am testing backend code [here](https://github.com/biocompute-objects/Galaxy-To-BCO/blob/dev/test_code/jsonParse-history-py3.py) and will integrate it in this repo once the following issues are addressed:
1) its written in python 3 because the initial function creates a blank BCO from a specified schema (will be more extensible in the future). Will need to re-write the functions in python2.7. 
2) writing in something that is local and I could test because I am unsure of the API and the backend code structure of galaxy
3) there are still a few items that do not match the BCO spec, mainly because I do not see exactly how we can pull that info automatically. there may need to be some manual entry. I am not sure how that could be best accomplished. Maybe by extending the options for history tags/annotation?

**Comment by corburn** [comment]
In regards to re-writing the functions in python2.7, galaxy is officially moving to Python 3:

>https://github.com/galaxyproject/galaxy/issues/1715#issuecomment-544551578
If I am making a new library for potential future use by Galaxy (say for exporting workspace as RO-Crate) - am I OK to do this in Python 3 or do I need to make it backwards compatible to Python 2? From the progress in here it sound like I can dare to be modern..! :)
>
>https://github.com/galaxyproject/galaxy/issues/1715#issuecomment-544572744
IMO go for python 3. We will have a deprecation notice in the upcoming release 7efbd04
and 19.09 should be the last release that officially supports python 2.


Galaxy Deprecation Notice:
>https://github.com/galaxyproject/galaxy/commit/7efbd043ede6e5fbb72bb445aa5d8a2cc6841fb5
Python2 will `not be maintained <https://www.python.org/doc/sunset-python-2/>`__ past 2020 which means Galaxy 19.09 is
the last release that officially supports Python 2.

**Comment by HadleyKing** [comment]
Does it make more sense to create a couple of functions and then to call them in the bco function or does it make more sense to have it all in one function?

**Comment by corburn** [comment]
I did not find a specific guideline in the best practices, but I'd follow the patterns of nearby functions. As a general rule, I would break it into separate functions if:
- the function is longer than a page
- it is hard to trace the variables or the logic is hard to follow

https://galaxyproject.org/develop/best-practices/
https://galaxy-iuc-standards.readthedocs.io/en/latest/

**Comment by corburn** [comment]
I pushed a `ci` branch that attempts to fill in the output_subdomain.

https://github.com/biocompute-objects/galaxy/tree/ci

**Comment by HadleyKing** [comment]
BCO has been moved to Workflows instead of Histories now. b038b7626181737383cf8e49edbd2279e51fb854
